### ISSUE TITLE
Collection View Cells on top of one another

### ISSUE THREAD

**Issue by DominikButz** [Digipolitan/collection-view-left-align-flow-layout-swift#3]
I have a UITableView with cells some of which have a UICollectionView as subview. All collection views have a cell type with one text label, except for one which has a UIImageView, no text label. Since photos can be portrait or landscape, the collection view cells can have different heights and widths. the collection view cell type with only a label only has varying widths, but all cells have the same height.
Now here is the problem: adding new cv cells of the one with the image view will put them all on top of each other instead of side by side, row by row.  

![screen shot 2017-07-12 at 14 04 00](https://user-images.githubusercontent.com/5699871/28104096-f0674bd8-670b-11e7-9bf7-d8296434a92a.png)

With the label type cv cell, that problem does not occur:
![screen shot 2017-07-12 at 14 04 26](https://user-images.githubusercontent.com/5699871/28104109-ff13c24c-670b-11e7-9db6-6053a6b698fd.png)

![screen shot 2017-07-12 at 14 04 48](https://user-images.githubusercontent.com/5699871/28104129-1c6ac26e-670c-11e7-9e15-4076c260b261.png)

I can't exclude I'm overlooking something  - but with the standard UICollectionViewFlowLayout the problem does not occur. 
thanks for any hints in the right direction

**Comment by julien-sarazin** [comment]
We will test it by the end of the week and come back to you ASAP. You might have spotted an issue regarding the height differences.

**Comment by DominikButz** [comment]
Thanks. I think the problem lies in this line:
let firstInRow = previousFrame.origin.y != attributes.frame.origin.y

If I add more cells, they actually get displayed in additional rows, too. But all cells in the same row have the same origin x...  that's why they appear clustered in the same place in each row. 

I have replaced it with this line 
   let firstInRow = (previousFrame.origin.y + previousFrame.height / 2.0)  !=  (attributes.frame.origin.y + attributes.frame.height / 2.0)

Despite the change, I still have trouble to make it work in my own app, I need to look into this further.

But it seems to work in the example project:

![screen shot 2017-07-23 at 14 09 26](https://user-images.githubusercontent.com/5699871/28497112-0d2bcd54-6fb1-11e7-9015-6602fe5fc60a.png)

**Comment by DominikButz** [comment]
There seems to be an issue if the heights are even and odd numbers, so the centre points might not always be the same for each row. therefore I changed the firstInRow definition again:

  let firstInRow = (previousFrame.origin.y + previousFrame.height / 2.0)  <  (attributes.frame.origin.y + attributes.frame.height / 2.0)  - 1.0

I have forked the repository and committed the change. 

I'm not 100% sure this works for all cases but I checked it for equal and unequal heights in your example.

**Comment by julien-sarazin** [comment]
You are right. To determine either or not the current attributes represent a new row, we are checking if the previous attributes origin (y) is different. This involve a massive side effect regarding the height. If an item has a different height from its previous neighbour it will be considered as a new row, meaning it will be placed at the most left position. I'm not sure you're formula is taking all cases into account, but it will definitely reduce the side effect. My team and I will try to fix this tomorrow.
Thanks again!

**Comment by julien-sarazin** [comment]
```swift
let firstInRow = previousAttributes?.center.y != attributes.center.y
```
Works perfectly for me for different sizes, or same sizes, or just different height. So far it covers all use cases.

**Comment by DominikButz** [comment]
Thanks, that is even simpler.  However, this doesn't work in my app. To calculate the size of each collection view cell in my app, the image is rescaled (keeping the width height ratio) to fit into a square of 100 x 100.  I logged the heights of the sizes that are returned in the collectionviewlayout delegate function  - e.g. 74.9063670411985   .  there might be a problem with rounding or decimal imprecision...   **This means that the centre y  values are not necessarily exactly the same among the cells of the same row.**  

Therefore I would suggest:

 let firstInRow = previousAttributes!.center.y < attributes.center.y - 1.0

The centre y value of the last cell in the previous row must be smaller than the centre y value of first cell in the current row. The question is by how much. I tried deducting 0.1, even that doesn't work. but in my tests, - 1.0 does work. 

![screen shot 2017-07-23 at 18 24 32](https://user-images.githubusercontent.com/5699871/28498636-48cf6442-6fd4-11e7-9bb1-05be7d9da480.png)
